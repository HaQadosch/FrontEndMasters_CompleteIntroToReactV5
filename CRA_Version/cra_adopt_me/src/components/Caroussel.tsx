import React, { useState, ReactEventHandler, useEffect } from 'react';
import { Photo } from '@frontendmasters/pet';

export const Caroussel: React.FC<{ photos: Photo[] }> = ({ photos: media }) => {
  const smallDefault = 'http://placecorgi.com/300/300';
  const largeDefault = 'http://placecorgi.com/600/600';
  const [photos, setPhotos] = useState<Photo[]>(media);
  const [active, setActive] = useState<number>(0);

  useEffect(() => {
    setPhotos(media.map(photo => ({ large: largeDefault, small: smallDefault, ...photo })));
  }, [media]);

  const onImgClick = (index: number): ReactEventHandler => ({ target }: React.SyntheticEvent) => {
    setActive(index);
  };

  return (
    <div className='carousel'>
      <img src={photos[active].large} alt='active pet' />
      <div className='carousel-smaller'>
        {photos.map((photo, index) => (
          <img
            key={index}
            src={photo.small}
            alt='small pet'
            className={index === active ? 'active' : ''}
            onClick={onImgClick(index)}
          />
        ))}
      </div>
    </div>
  );
};
