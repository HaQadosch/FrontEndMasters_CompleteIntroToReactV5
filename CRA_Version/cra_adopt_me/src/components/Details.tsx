import React, { useEffect, useState, useReducer, Dispatch } from 'react';
import pet, { Animal } from '@frontendmasters/pet';

const animalReducer = (currentState: any, newState: any) => {
  return { ...currentState, ...newState };
};

export const Details: React.FC<{ path: string; id?: string }> = ({ id = '' }) => {
  const [loading, setLoading] = useState<Boolean>(true);
  const [, /*error*/ setError] = useState<Error>();

  const [
    {
      name,
      type: animal,
      contact: {
        address: { city, state },
      },
      description,
      /*photos: media,*/ breeds: { primary: breed },
    },
    setState,
  ]: [Animal, Dispatch<any>] = useReducer(animalReducer, {
    name: '',
    animal: '',
    description: '',
    media: [],
    contact: { address: { city: '', state: '' } },
    breeds: { primary: '' },
  });

  useEffect(() => {
    pet
      .animal(parseInt(id))
      .then(({ animal }) => {
        setState(animal);
        setLoading(false);
      })
      .catch(petErr => setError(petErr));
  }, [id]);

  return loading ? (
    <h1>
      <span role='img' aria-label='sand-glass'>
        âŒ›
      </span>
      Loading...
    </h1>
  ) : (
    <div className='details'>
      <div>
        <h1>{name}</h1>
        <h2>{`${animal} - ${breed} - ${city}, ${state}`}</h2>
        <button>{`Adopt ${name}`}</button>
        <p>{description}</p>
      </div>
    </div>
  );
};

export default Details;
