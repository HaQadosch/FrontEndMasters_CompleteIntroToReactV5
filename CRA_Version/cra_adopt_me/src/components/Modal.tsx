import React, { useRef, useEffect } from 'react';
import { createPortal } from 'react-dom';

export const Modal: React.FC = ({ children }) => {
  const elt = useRef<HTMLDivElement>();
  if (!elt.current) {
    const div = document.createElement('div');
    elt.current = div;
  }

  useEffect(() => {
    const modalRoot = document.querySelector('#modal');
    if (modalRoot && elt.current) {
      modalRoot.appendChild(elt.current);
    }
    return () => {
      if (modalRoot && elt.current) {
        modalRoot.removeChild(elt.current);
      }
    };
  }, []);

  return createPortal(<div>{children}</div>, elt.current);
};
