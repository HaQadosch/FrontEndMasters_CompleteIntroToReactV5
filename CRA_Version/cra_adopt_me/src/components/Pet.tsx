import React from 'react';
import { Animal } from '@frontendmasters/pet';
import { Link } from '@reach/router';

export const Pet: React.FC<Animal> = ({
  id,
  name,
  type: animal,
  breeds: { primary: breed } = { primary: '' },
  photos: media = [],
  contact: { address } = { address: { city: '', state: '' } },
}) => {
  let hero = media.length ? media[0].small : 'http://placecorgi.com/300/300';
  return (
    <Link to={`/details/${id}`} className='pet'>
      <div className='image-container'>
        <img src={hero} alt={name} />
      </div>
      <div className='info'>
        <h1>{name}</h1>
        <h2>{`${animal} - ${breed} - ${address.city}, ${address.state}`}</h2>
      </div>
    </Link>
  );
};
