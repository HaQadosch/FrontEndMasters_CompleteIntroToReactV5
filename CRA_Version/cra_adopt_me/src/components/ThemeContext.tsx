import React, { createContext, SetStateAction, Dispatch } from 'react';

// export const ThemeContext: React.Context<(string | (() => string))[]> = createContext(['green', () => ''])
export const ThemeContext: React.Context<(string | (Dispatch<SetStateAction<string>>))[]> = createContext([
  'green',
  (_: SetStateAction<string>) => {},
]);
