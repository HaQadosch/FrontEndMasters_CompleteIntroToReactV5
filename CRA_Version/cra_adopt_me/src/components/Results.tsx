import React from 'react';
import { Pet } from './Pet';
import { Animal } from '@frontendmasters/pet';

export const Results: React.FC<{ pets: Animal[] }> = ({ pets }) => {
  return (
    <div className='search'>
      {pets.length === 0 ? <h1>No pets found.</h1> : pets.map(({ id, ...pet }) => <Pet key={id} id={id} {...pet} />)}
    </div>
  );
};
