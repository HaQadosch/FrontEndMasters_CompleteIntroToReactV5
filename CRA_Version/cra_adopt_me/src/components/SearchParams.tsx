import React, { useState, useEffect, useContext } from 'react';
import pet, { ANIMALS, AnimalsResponse, Animal, AnimalsParams } from '@frontendmasters/pet';
import { Results } from './Results';
import { useDropdown } from '../utils/useDropdown';
import { ThemeContext } from './ThemeContext';

export const SearchParams: React.FC<{ path: string }> = () => {
  const [location, setLocation] = useState<string>('Seattle, WA');
  const [breeds, setBreeds] = useState<string[]>([]);
  const [animal /*setAnimal*/, , AnimalDropdown] = useDropdown('Animal', 'all', ANIMALS);
  const [breed, setBreed, BreedDropdown] = useDropdown('Breed', 'dog', breeds);
  const [pets, setPets] = useState<Animal[]>([]);
  const [theme, setTheme] = useContext<(string | any)[]>(ThemeContext);

  const requestPet = async (requestParams: AnimalsParams) => {
    const { animals }: AnimalsResponse = await pet.animals(requestParams);
    setPets(animals || []);
  };

  useEffect(() => {
    setBreeds([]);
    setBreed('');

    pet.breeds(animal).then(({ breeds: petBreeds }) => {
      if (petBreeds) {
        const breedsNames = petBreeds.map(({ name }) => name);
        setBreeds(breedsNames);
      }
    }, console.error);
  }, [animal, setBreed]);

  const onInputLocationChange: React.ChangeEventHandler<HTMLInputElement> = ({ target: { value } }) =>
    setLocation(value);

  const onFormSubmit: React.ChangeEventHandler<HTMLFormElement> = evt => {
    evt.preventDefault();
    requestPet({
      location,
      breed,
      type: animal,
    });
  };

  const onThemeSelectChange: React.ChangeEventHandler<HTMLSelectElement> = ({ target: { value } }) => {
    setTheme(value);
  };

  return (
    <div className='search-params'>
      <form action='/' onSubmit={onFormSubmit}>
        <label htmlFor='location'>
          Location
          <input
            type='text'
            name='location'
            id='location'
            value={location}
            onChange={onInputLocationChange}
            placeholder='location'
          />
        </label>
        <AnimalDropdown />
        <BreedDropdown />
        <label htmlFor='theme'>
          Theme
          <select name='theme' id='theme' value={theme} onChange={onThemeSelectChange} onBlur={onThemeSelectChange}>
            <option value='pere'>peru</option>
            <option value='#bada55'>bada55</option>
            <option value='mediumorchid'>medium orchid</option>
          </select>
        </label>
        <button type='submit' style={{ backgroundColor: theme }}>
          Submit
        </button>
      </form>
      <Results pets={pets} />
    </div>
  );
};
