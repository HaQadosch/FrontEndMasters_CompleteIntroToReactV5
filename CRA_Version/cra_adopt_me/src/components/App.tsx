import React, { useState } from 'react';
import { Router, Link } from '@reach/router';
import './App.css';
import { SearchParams } from './SearchParams';
import { Details } from './Details';
import { ThemeContext } from './ThemeContext';

export const App: React.FC = () => {
  const [color, setColor] = useState<string>('darkblue');

  return (
    <ThemeContext.Provider value={[color, setColor]}>
      <div>
        <header>
          <Link to={'/'}>Adopt me!</Link>
        </header>
        <React.StrictMode>
          <Router>
            <SearchParams path='/' />
            <Details path='/details/:id' id='' />
          </Router>
        </React.StrictMode>
      </div>
    </ThemeContext.Provider>
  );
};
