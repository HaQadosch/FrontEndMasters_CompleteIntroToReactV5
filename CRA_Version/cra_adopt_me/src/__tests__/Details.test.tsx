import React from 'react';
import ReactDOM from 'react-dom';
import { render } from 'react-testing-library';
import { Details } from '../components/Details';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Details />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('Details matches the snapshot', () => {
  const { debug, asFragment } = render(<Details />);
  expect(asFragment()).toMatchSnapshot();
  // debug(asFragment());
});
