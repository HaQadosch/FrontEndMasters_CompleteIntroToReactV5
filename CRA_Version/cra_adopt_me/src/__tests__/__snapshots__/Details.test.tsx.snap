// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details matches the snapshot 1`] = `
<DocumentFragment>
  <h1>
    <span
      aria-label="sand-glass"
      role="img"
    >
      âŒ›
    </span>
    Loading...
  </h1>
</DocumentFragment>
`;
